BUG: Add --read_arguments_from_file to sl_fastq.py

ElDeveloper opened the pull request:
With this new argument split_libraries_fastq.py can now take a file that
list the input files for the `-b`, `-i`, `--sample_ids` and
`--mapping_files` arguments. Thus allowing an indefinite amount of files
as inputs, as opposed to the previous approach where we would be limited
by the maximum number of characters that can fit into a single command
line execution.

Fixes #2069

cc @walterst @gregcaporaso

qiime-r2d2 commented:
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1668/

qiime-r2d2 commented:
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1669/

ElDeveloper commented:
BTW, the tests are passing, the only thing failing at this point is `test_otu_significance.py` but it seems like that test was already failing, see #2078.

Could I get someone to review this? I'm specifically concerned that this new argument can be confusing to users and that maybe it should be "hidden"??

antgonza commented:
The code looks fine and the test failure is something we already know about. Thus: :+1: However, I think @walterst or @gregcaporaso should take a look also.

walterst commented:
One comment about the code that @ElDeveloper has added-do we want the temporary file(s) that are created to be in the current directory rather than the /tmp/ (or whatever the temp folder is set to) directory? It may be more apparent (particularly for users troubleshooting issues that may come up with the generated --sample_id names) if the files created for them are in the current output directory. This would require an added call to create the output directory (normally created with the split_libraries_fastq.py command is called) before the temp files are created.

ElDeveloper commented:
Thanks both.

@walterst, as a matter of fact `multiple_split_libraries_fastq.py` is already creating the output directory, so this makes a lot of sense.

ElDeveloper commented:
@walterst I just added that, can you have a loo again? I've also expanded the documentation to mention these files.

ElDeveloper commented:
BTW in the process of working in the initial submission for this PR, I found a bug where "trailing text" would not really be the "trailing text", see [this line](https://github.com/biocore/qiime/blob/master/tests/test_workflow/test_preprocess.py#L358), specifically see how '--barcode-type' is at the very end of the line, before the trailing text.

adamrp commented on qiime/workflow/preprocess.py:
```diff
@@ -121,14 +124,27 @@ def create_commands_slf(all_files, demultiplexing_method, output_dir,
     sample_ids = []
 
+    params_dict = {}
+    path_builder = lambda x: join(output_dir, x)
```
`functools.partial`?
```python
path_builder = partial(join, output_dir)
```

ElDeveloper replied:
Hmmm, I don't think I can do that, if I were to do that then output_dir
would be appended to the name of my file i.e. path_builder('foo.hdf5')
would result in 'foo.hdf5/output_dir_name'.

On (Aug-28-15|16:20), adamrp wrote:
> >      commands = []
> >      read_files = []
> >      barcode_files = []
> >      mapping_files = []
> >      sample_ids = []
> >  
> > +    params_dict = {}
> > +    path_builder = lambda x: join(output_dir, x)
> 
> `functools.partial`?
> ```python
> path_builder = partial(join, output_dir)
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077/files#r38253638

adamrp replied:
I don't think that's accurate:
```python
>>> from functools import partial
>>> from os.path import join
>>> output_dir = 'some_directory'
>>> path_builder = partial(join, output_dir)
>>> filename = 'foo.hdf5'
>>> the_path = path_builder(filename)
>>> the_path
'some_directory/foo.hdf5'
>>> 
```

adamrp commented on qiime/workflow/preprocess.py:
```diff
@@ -148,22 +164,56 @@ def create_commands_slf(all_files, demultiplexing_method, output_dir,
+
+    Parameters
+    ----------
+    elements : list
```
more generally, could the type be any `iterable`?

ElDeveloper replied:
Yes, though in the context of create_commands_slf, it should only take
ordered iterables, thus my decision to make it `list`. Otherwise
sequences and sample names could get out of sync :scream: :scream:
What about changing this to list, tuple?

On (Aug-28-15|16:26), adamrp wrote:
> >  
> >      return commands
> >  
> > +
> > +def _make_file(elements, fp):
> > +    """Create a file with one element per line
> > +
> > +    This function is just a helper for create_commands_slf.
> > +
> > +    Parameters
> > +    ----------
> > +    elements : list
> 
> more generally, could the type be any `iterable`?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077/files#r38253935

adamrp replied:
it is a private function, so this is probably too nitpicky :smile: I'm fine leaving at as `list`, but if you wanted to change it to be more general, that would be fine, too.

adamrp replied:
And I do take your point about ordering, that *is* important

adamrp commented on scripts/split_libraries_fastq.py:
```diff
@@ -151,6 +161,24 @@
+    # allow arguments that would span over hundreds of samples and would be
+    # prohibitive to execute as a command line call
+    if read_arguments_from_file:
+        read_f = lambda x: open(x, 'U').read().strip().split('\n')
```
would it be better to actually create this function using `def` so that you can use a context manager with `open`?

adamrp commented on tests/test_workflow/test_preprocess.py:
```diff
@@ -315,14 +315,18 @@ def test_create_commands_slf_by_sampleID(self):
+                    "'not-barcoded' ").format(output_dir=output_dir)
 
-        self.assertEqual(actual_command, expected_command)
+        self.assertEquals(actual_command, expected)
```
`assertEquals` is a [deprecated synonym](https://docs.python.org/2/library/unittest.html#deprecated-aliases), better to use `assertEqual`

adamrp commented:
I reviewed the code, but I won't have time to test it until later tonight. Just a few minor comments, otherwise code looks good to me.

ElDeveloper commented:
Thanks a lot @adamrp, I think I've addressed all your comments. Pretty
helpful!!

On (Aug-28-15|16:58), adamrp wrote:
> I reviewed the code, but I won't have time to test it until later tonight. Just a few minor comments, otherwise code looks good to me.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077#issuecomment-135917672

qiime-r2d2 commented:
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1672/

adamrp commented:
:+1:

antgonza commented:
This is ready to go but I think we should wait to solve the test_otu_significance.py before merging. What do you guys think?

ElDeveloper commented:
Sounds good, though I would prefer we didn't wait indefinitely. What do
you guys think?

On (Aug-29-15| 6:51), Antonio Gonzalez wrote:
> This is ready to go but I think we should wait to solve the test_otu_significance.py before merging. What do you guys think? 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077#issuecomment-135988438

gregcaporaso commented:
@ElDeveloper, do you still need additional review of this? If so, let me know and I'll take a pass.

ElDeveloper commented:
@gregcaporaso, yes that would be fantastic. Also if you could give some
input on whether you also believe we should hold merge of this PR before
we resolve the test failure in test_otu_significance.py (see #2078).

On (Sep-01-15| 9:47), Greg Caporaso wrote:
> @ElDeveloper, do you still need additional review of this? If so, let me know and I'll take a pass. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077#issuecomment-136789259

ElDeveloper commented:
Just as an update, I followed up win #2078 but there doesn't seem to be a clear/quick way to get those tests passing again, master is also failing, so I would suggest that we merge this (granted that all reviewers agree).

antgonza commented:
My worry was master but as now is also failing, I'm gonna merge.

ElDeveloper commented:
Thanks @antgonza, @walterst and @adamrp!!
